Add devices argument to RRef constructor

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57085 Add devices argument to RRef constructor**

PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

mrshenli committed:
Add devices argument to RRef constructor

PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

mrshenli commented:
I will add one more PR on top to make `request_callback_impl.cpp` to use `CUDAFuture` too

mrshenli commented:
Will wait for the generic `MultiStreamGuard` in #57049 before updating `request_callback_impl.cpp `

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

lw reviewed (approved):
This code looks good to me, but I know very little about RRefs so you could consider having someone else take a look too.

I think it's fine to ignore/silence the lint errors, since that code will go away once CUDAFuture is merged into ivalue::Future.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -206,9 +223,13 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+                  const py::object&,
+                  optional<std::vector<c10::DeviceIndex>>>(),
               py::arg("value"),
-              py::arg("type_hint") = py::none())
+              py::arg("type_hint") = py::none(),
```
Looks yes. I will only do this for `devices` for now, as `type_hint` already allow position arguments since a few releases ago.

https://pybind11.readthedocs.io/en/stable/advanced/functions.html#keyword-only-arguments

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -206,9 +223,13 @@ PyObject* rpc_init(PyObject* _unused, PyObject* noargs) {
+              py::init<
+                  const py::object&,
+                  const py::object&,
+                  optional<std::vector<c10::DeviceIndex>>>(),
```
Yep, when reading your PR, I was thinking about the same thing. Will do that on top of this PR. It should be fine to only accept int for now and then generalize to int/string/device, as there won't be BC issue.

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]

mrshenli committed:
Update on "Add devices argument to RRef constructor"


PR #54932 fixed the CUDA RPC for RRef when RRef is created through
RPC. But besides that use case, RRef can also be created locally
by directly passing in a value, which would bypass the CUDA stream
synchronization in #54932.

This commit covers the above gap by adding a `devices` argument
to RRef constructor. The RRef will then use this argument to
choose between `CUDAFutre` and `ivalue::Future` to hold the value.
When `devices` is specified and non-empty, `CUDAFuture` will be
used, and the `devices` will be passed to that `CUDAFuture`.

Differential Revision: [D28050001](https://our.internmc.facebook.com/intern/diff/D28050001)

[ghstack-poisoned]
